From d84bbb2308ce904c43693fcea271d61f590136e7 Mon Sep 17 00:00:00 2001
From: WofWca <wofwca@protonmail.com>
Date: Thu, 17 Sep 2026 16:40:01 +0400
Subject: [PATCH] fix: random gibs bounciness: same seed for all

This actually also increases the rotation damping,
syncing it with linear damping better.
---
 code/cgame/cg_effects.c   | 10 ++++++++--
 code/cgame/cg_localents.c | 10 ++++++----
 2 files changed, 14 insertions(+), 6 deletions(-)

diff --git a/code/cgame/cg_effects.c b/code/cgame/cg_effects.c
index eb9f7d8bc..fd4584e81 100644
--- a/code/cgame/cg_effects.c
+++ b/code/cgame/cg_effects.c
@@ -533,6 +533,7 @@ void CG_LaunchGib( const vec3_t origin, const vec3_t angles,
 					const int randSeed ) {
 	localEntity_t	*le;
 	refEntity_t		*re;
+	int seed = randSeed;
 
 	le = CG_AllocLocalEntity();
 	re = &le->refEntity;
@@ -550,14 +551,19 @@ void CG_LaunchGib( const vec3_t origin, const vec3_t angles,
 	VectorCopy( velocity, le->pos.trDelta );
 	le->pos.trTime = cg.time;
 
-	le->bounceFactor = cg_oldGibs.integer ? 0.6f : cg_gibsBounceFactor.value;
+	le->bounceFactor = 0.6f;
+	if ( !cg_oldGibs.integer ) {
+		// Same calculation as in `CG_ReflectVelocity`.
+		float r = ( Q_random(&seed) + Q_random(&seed) ) / 2;
+		le->bounceFactor = cg_gibsBounceFactor.value;
+		le->bounceFactor *= 1 - r * cg_gibsBounceFactorRandomness.value;
+	}
 
 	if (!cg_oldGibs.integer) {
 		// `VectorLength` would be more precise, but this is faster
 		// and good enough for randomness.
 		float speedIsh = fabs(velocity[0]) + fabs(velocity[1]) + fabs(velocity[2]);
 		int i;
-		int seed = randSeed;
 		int mainRotationAxis = Q_rand(&seed) % 3;
 
 		le->leFlags = LEF_TUMBLE;
diff --git a/code/cgame/cg_localents.c b/code/cgame/cg_localents.c
index 7801a1081..dabf7e75d 100644
--- a/code/cgame/cg_localents.c
+++ b/code/cgame/cg_localents.c
@@ -220,7 +220,6 @@ void CG_ReflectVelocity( localEntity_t *le, trace_t *trace, vec3_t velocityDiffe
 	vec3_t	velocity;
 	float	dot;
 	int		hitTime;
-	float	bounceFactor = le->bounceFactor;
 
 	// reflect the velocity on the trace plane
 	hitTime = cg.time - cg.frametime + cg.frametime * trace->fraction;
@@ -229,11 +228,14 @@ void CG_ReflectVelocity( localEntity_t *le, trace_t *trace, vec3_t velocityDiffe
 	VectorMA( velocity, -2*dot, trace->plane.normal, le->pos.trDelta );
 
 	if ( le->leBounceSoundType == LEBS_BLOOD && !cg_oldGibs.value ) {
+		int seed = le->bounceFactor * 0x1000000;
 		// Get a tighter distribution rather than uniform.
-		float r = ( random() + random() ) / 2;
-		bounceFactor *= 1 - r * cg_gibsBounceFactorRandomness.value;
+		// Same calculation as in `CG_LaunchGib`.
+		float r = ( Q_random(&seed) + Q_random(&seed) ) / 2;
+		le->bounceFactor = cg_gibsBounceFactor.value;
+		le->bounceFactor *= 1 - r * cg_gibsBounceFactorRandomness.value;
 	}
-	VectorScale( le->pos.trDelta, bounceFactor, le->pos.trDelta );
+	VectorScale( le->pos.trDelta, le->bounceFactor, le->pos.trDelta );
 
 	if (velocityDifference) {
 		VectorSubtract( le->pos.trDelta, velocity, velocityDifference );
-- 
2.56.0

